### PR TITLE
[Session] limiting :key for GET_LOCK to 64 chars

### PR DESCRIPTION
**Pull request by oleg-andreyev**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT

> MySQL 5.7.5 and later enforces a maximum length on lock names of 64 characters. Previously, no limit was enforced.

Cases:
- `session_id` is set by developers manually
- `session.sid_length` is configured

Ref.: 
- https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_get-lock
- http://php.net/manual/en/session.configuration.php#ini.session.sid-length

Other issues:
- https://github.com/go-sql-driver/mysql/issues/385
- https://github.com/stefangabos/Zebra_Session/issues/16

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 1]
```diff
@@ -674,14 +674,17 @@ private function doAdvisoryLock(string $sessionId)
     {
         switch ($this->driver) {
             case 'mysql':
+                // MySQL 5.7.5 and later enforces a maximum length on lock names of 64 characters. Previously, no limit was enforced.
+                $lockId = \substr($sessionId, 0, 64);
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: I'd recommend to hash the session id instead (or doing something similar) as keeping just the first 64 chars could lead to conflicts depending on how the session id is generated.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php (hunk 1 above)
Comment (Tobion) [reply src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: The session id is just a random string. So hashing that does not reduce the likeliness of collisions compared to just using the first 64 bytes.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 2]
```diff
@@ -833,7 +836,7 @@ private function getUpdateStatement($sessionId, $sessionData, $maxlifetime)
     /**
      * Returns a merge/upsert (i.e. insert or update) statement when supported by the database for writing session data.
      */
-    private function getMergeStatement(string $sessionId, string $data, int$maxlifetime): ?\PDOStatement
+    private function getMergeStatement(string $sessionId, string $data, int $maxlifetime): ?\PDOStatement
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: This has already been fixed in https://github.com/symfony/symfony/commit/4f3afd53a8e3c673b889eb936410291108c89d88 and will be merge upwards in later branches automatically

**Comment by Tobion** [comment]
To me this is a bugfix and should be opened against 2.7 branch.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php (hunk 1 above)
Comment (oleg-andreyev) [reply src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: @fabian, @Tobion is right, session_id is just a random string, which is result of http://php.net/manual/en/session.configuration.php#ini.session.hash-function and by default it's md5

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php (hunk 2 above)
Comment (oleg-andreyev) [reply src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: Will rebase branch

**Comment by oleg-andreyev** [comment]
@Tobion agree, let's rebase on 2.7

**Review by nicolas-grekas — APPROVED** [review]

**Review by Tobion — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @oleg-andreyev.

**Pull request merged**